### ISSUE TITLE
High failure rate reading

### ISSUE THREAD

**Issue by gonsp** [MarcProe/LibrePilot.arduino#1]
I've tried it (good work!) in two different flight controllers (CC3D & Revo) and with an Arduino Uno and an Arduini Mini Pro.

I'm getting a really high failure rate reading UAVObjects from the flightcontroller. If fails like 6/10 requests.

Is it normal? It's a very big failure rate considering it's a wired serial connection... Maybe it's the use of the SoftwareSerial which doesn't perform correctly? What failure rate do you have?

Thanks!

**Comment by MarcProe** [comment]
What exactly do you mean by "high failure rate"?

Currently, this code listens to the UAVO-Stream and returns if the object you are looking for is matched. If the object is not send, the function returns after the timeout.
You might want to try and set a higher timeout.

**Comment by gonsp** [comment]
Sorry for not having answered before.

I've tried higher timeouts and different Arduinos and I still get a high fail rate. Most of the times it's due to the crc.

But I've seen that my arduinos are 5V logical and the serial output of the revo is 3V (5V tolerant). So maybe that was the problem. I've read the specs from Arduino and questions in SO and theoretically that doesn't matter, but it's the only reason I can think of.
I also tried to use the "hardware" serial instead of SoftwareSerial to see if it works better but it doesn't.

I'll close the issue since I think it's the problem with the voltage levels.

Thanks!

**Comment by jubaleth** [comment]
Hi!

I've also noticed this issue, and can confirm that the voltage is not the problem in this case. I found this issue, and decided to put in a level shift on the Tx and Rx.

I was originally seeing around 67% failed serial requests (so object request and then subsequent read). I managed to get this down to about 18-20% by adding a 3ms delay between requesting an object, and trying to receive it. 

What I've observed now is that there are a lot of reads where the sync byte is not 0x3c.

![image](https://user-images.githubusercontent.com/1205712/43036164-29b387c8-8cfc-11e8-96ed-b7388d6c273e.png)

and as @GonMolon said, there are a lot of invalid CRCs happening.

![image](https://user-images.githubusercontent.com/1205712/43036177-97fb94f0-8cfc-11e8-84ec-cc08addbdf7f.png)
 
I'm going to try to find some time to sit down with a logic analyzer and see what's happening on the wire.
